### PR TITLE
Bug fixes & cache merged gateways to optimize RDS

### PR DESCRIPTION
**Pull request by rshriram**
This is an alternate way of doing https://github.com/istio/istio/pull/16688
that spotted a serious perfomance issue when the gateway had a lot of routes.

The merged gateways for a workload are precomputed before a push cycle.
Along the way I have also fixed a serious bug: the node.WorkloadLabels are
not being used uniformly. Several places in gateway code (and in some sidecar listener code)
were still using the labels from the proxy service instances and not the ones populated
in the node.WorkloadLabels. And in a scenario where the proxy instances are yet to be populated,
Listener code or gateway code wont work.

Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -1136,13 +1137,14 @@ func (ps *PushContext) EnvoyFilters(proxy *Proxy) []*EnvoyFilterWrapper {
 
 // pre computes gateways per namespace
 func (ps *PushContext) initGateways(env *Environment) error {
-	gatewayConfigs, err := env.List(schemas.Gateway.Type, NamespaceAll)
+	gatewayConfigs, err := env.IstioConfigStore.List(schemas.Gateway.Type, NamespaceAll)
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go]: what is the impact of this change?

Comment (rshriram) [reply pilot/pkg/model/push_context.go]: aren't they the same? One of the tests was segfaulting if the istioConfigStore was not used. So I thought it was something to do with how the fake store is setup.

Comment (howardjohn) [reply pilot/pkg/model/push_context.go]: Not sure.. probably something that can/should be cleaned up. Seems most places we just use `env.List`

**Comment by howardjohn** [comment]
We need ISTIO_METAJSON_LABELS in the gateways I think. Maybe won't solve the problem though

**Pull request merged**